### PR TITLE
Add indirection to FileWatchers.

### PR DESCRIPTION
**Pull request by aemoncannon**
Setting the stage for an optional Java 7 based watcher.

### CODE REVIEW

**Comment by fommil** [comment]
nice! Hopefully one of us will get there over the coming weeks.

I can set up an ensime-docker 2.x image with OpenJDK 7 on it, so we can build a standalone project that implements the interface. Ideally, matching the VFS2 API too, but not necessarily. I'd also like to fix the fundamental bugs in the VFS file watcher implementation anyway.

BTW, you failed the formatting test.

**Comment by fommil** [comment]
sorry I think I restarted the servers while this build was going on

**Comment by fommil** [comment]
I'll have a look at this in the next day or two, been a busy start to the year.

**Comment by fommil** [comment]
I'll review this first thing in the morning, getting the file watchers working on big windows projects is going to really help me

File: core/src/main/scala/org/ensime/indexer/FileWatchers.scala [hunk 1]
```diff
@@ -14,16 +14,14 @@ import scala.concurrent.Future
 import scala.util.Try
 import org.ensime.util.file._
 
-trait ClassfileListener {
-  def classfileAdded(f: FileObject): Unit
-  def classfileRemoved(f: FileObject): Unit
-  def classfileChanged(f: FileObject): Unit
+trait FileChangeListener {
```
Comment (fommil) [inline core/src/main/scala/org/ensime/indexer/FileWatchers.scala]: I should never have created two traits in the first place.

Comment (fommil) [inline core/src/main/scala/org/ensime/indexer/FileWatchers.scala]: maybe I was trying to give myself the false illusion of type safety.

File: core/src/main/scala/org/ensime/indexer/FileWatchers.scala [hunk 2]
```diff
@@ -79,10 +144,10 @@ class ClassfileWatcher(
         running = scheduler.schedule(resetter, 1, TimeUnit.SECONDS)
       }
 
-      private val targets: Set[FileName] = for {
-        dir <- config.targetClasspath
+      private val targets: Set[FileName] = (for {
+        dir <- watchedDirs
         ref = vfs.vfile(dir)
-      } yield ref.getName()
+      } yield ref.getName()).toSet
```
Comment (fommil) [inline core/src/main/scala/org/ensime/indexer/FileWatchers.scala]: it should be slightly more efficient to take a Set[File] as watchedDirs (avoids an intermediate collection)

**Comment by fommil** [comment]
great, LGTM! If you could just change the `watchedDirs` to a `Set[File]` that will give us a small performance boost, then just merge when it turns green.

File: core/src/main/scala/org/ensime/indexer/FileWatchers.scala (hunk 2 above)
Comment (rorygraves) [inline core/src/main/scala/org/ensime/indexer/FileWatchers.scala]: Breakout......  Does nobody watch my talks :(

Comment (fommil) [inline core/src/main/scala/org/ensime/indexer/FileWatchers.scala]: no need! It is actually a `Set[File]` to begin with, but the type sig is `Iterable[Set]` so the compiler doesn't know that.

**Comment by fommil** [comment]
the best thing about this PR is that @aemoncannon's name is going to be all over the file watcher code now, not mine, so the evidence has been destroyed! :laughing:

**Comment by rorygraves** [comment]
You have been marked, it will always be your fault ;)

**Comment by fommil** [comment]
nope! look at the blame log. It's A E M O N . Not Sam, Aemon. It's obviously all his fault. I would never write such shoddy file watcher code. (squashes history, force pushes...)

**Pull request merged**